connect4.hs

```diff
@@ -0,0 +1,62 @@
+module Connect4 where
+
+import System.IO
+
+data State = State InternalState [Action]  -- internal_state available_actions
+         deriving (Eq, Show)
+
+data Result = EndOfGame Double State    -- end of game: value, starting state
+            | ContinueGame State        -- continue with new state
+         deriving (Eq, Show)
+
+type Game = Action -> State -> Result
+
+type Player = State -> Action
+
+------ Connect4 Game -------
+data Action = Action Int                 -- a move for a player is just an Int
+         deriving (Ord,Eq)
+
+data TeamColour = Red
+                | Black
+type InternalState = (Int, TeamColour, [[TeamColour]])   -- (open slots remaining, current colour's turn, 2d array of BoardSpace)
```

Comment:
shouldn't open slots remaining be a list of Ints?

Reply:
I meant number of slots (out of 42) to detect when theirs no more spots. We could do this differently if we say want to keep track of how many slots in each column